### PR TITLE
add exception handling around Activator.CreateInstance calls

### PR DESCRIPTION
**Pull request by davidkline-ms**
Overview
---
Adds exception handling around usage of Activator.CreateInstance to prevent the premature termination of MixedRealityToolkit.Initialize().

Changes
---
- Fixes: #3037 .

### CODE REVIEW

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 1]
```diff
@@ -174,8 +174,16 @@ private void InitializeServiceLocator()
...
-                if (!RegisterService(typeof(IMixedRealityInputSystem), Activator.CreateInstance(ActiveProfile.InputSystemType) as IMixedRealityInputSystem) ||
-                    InputSystem == null)
+                bool registrationSucceeded = false;
+                try
+                {
+                    registrationSucceeded = RegisterService(typeof(IMixedRealityInputSystem), Activator.CreateInstance(ActiveProfile.InputSystemType) as IMixedRealityInputSystem);
+                }
+                catch (Exception e)
+                {
+                    DebugUtilities.LogCriticalError($"Registering the Input System threw: {e.GetType().ToString()} - {e.Message}.");
```
Comment (ryzngard) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: Should we add `DebugUtilities.LogCriticalError(string message, Exception e)` to have common formatting?

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 2]
```diff
@@ -174,8 +174,16 @@ private void InitializeServiceLocator()
                 Utilities.Editor.InputMappingAxisUtility.CheckUnityInputManagerMappings(Definitions.Devices.ControllerMappingLibrary.UnityInputManagerAxes);
 #endif
 
-                if (!RegisterService(typeof(IMixedRealityInputSystem), Activator.CreateInstance(ActiveProfile.InputSystemType) as IMixedRealityInputSystem) ||
-                    InputSystem == null)
+                bool registrationSucceeded = false;
+                try
+                {
+                    registrationSucceeded = RegisterService(typeof(IMixedRealityInputSystem), Activator.CreateInstance(ActiveProfile.InputSystemType) as IMixedRealityInputSystem);
```
Comment (ryzngard) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: What if we had `RegisterService<T>(Func<T> factory)` and simplified this across all of them?

Comment (ryzngard) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: Callers would do `RegisterService<IMixedRealityInputSystem>(() => Activator.CreateInstance(…))`

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs (hunk 1 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: maybe. Not a bad thought

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs (hunk 2 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: I had thought of that as a possible optimization.

**Comment by StephenHodgson** [comment]
Agreed

**Comment by SimonDarksideJ** [comment]
This will cause conflicts with your new fix PR @davidkline-ms #3164

**Comment by davidkline-ms** [comment]
Merged revert of debug log change. Output from test runner shown below:

capture

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 3]
```diff
@@ -557,30 +552,78 @@ public bool RegisterService(Type type, IMixedRealityService service)
...
+
+            if (concreteType == null)
+            {
+                Debug.LogError("Unable to register a service with a null concrete type.");
+                return false;
+            }
+
+            if (!typeof(IMixedRealityService).IsAssignableFrom(concreteType))
+            {
+                Debug.LogError($"Unable to register the {concreteType.GetType()} service. It does not implement the IMixedRealityToolkitService interface.");
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: The interface checked above is `IMixedRealityService`, but this comment references `IMixedRealityToolkitService`. I think the comment needs to be updated.

**Comment by StephenHodgson** [comment]
Looks good, just [a few CRs](https://github.com/davidkline-ms/MixedRealityToolkit-Unity/pull/5)

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs (hunk 3 above)
Comment (keveleigh) [inline Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]:
```suggestion
                Debug.LogError($"Unable to register the {concreteType.GetType()} service. It does not implement the IMixedRealityService interface.");
```

File: Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs [hunk 4]
```diff
@@ -229,12 +224,10 @@ private void InitializeServiceLocator()
                     if (Application.platform.IsPlatformSupported(configuration.RuntimePlatform))
 #endif
                     {
-                        if (configuration.ComponentType.Type != null)
+                        if (configuration.ComponentType.Type != null &&
```
[the remark this answers is not in the document]
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: that feels outside of the scope of this change as the original code had the != null (do not want to fundamentally change behavior here).

[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: Seems reasonable as it checked if the service is registered correctly.

If the type is null then we should at least be notified.

[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: In the internal registration we're already checking if the type is null.

I'm wondering if we even need this check anyway.

[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Services/MixedRealityToolkit.cs]: Yup. Confirmed. Redundant check can go away completely.

**Pull request merged**